### ISSUE TITLE
Add a few high-level analysis benchmarks

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#2275]
In 2017 we had an issue with Gammapy where memory usage would grow with each observation processed due  to a coding mistake (should have stayed constant) and users complained that they ran out of memory.

We set up https://github.com/gammapy/gammapy-benchmarks and added one benchmark (see [iact_basic_image_estimator.py](https://github.com/gammapy/gammapy-benchmarks/blob/master/benchmarks/iact_basic_image_estimator.py) with results [here](https://github.com/gammapy/gammapy-benchmarks/blob/master/results/iact-basic-image-estimator/0.7.dev5319/2017-12-01/results.png). By now that script is broken of course.

In addition we have a dozen tutorial notebooks, and we can see how long they take to execute in the logs: https://travis-ci.org/gammapy/gammapy/jobs/554290538#L1268-L1339

But really we don't have any insight or overview over Gammapy performance concerning CPU and memory usage at the moment.

I suggest we spend some time on this, and pre-script the 5-10 most common analysis cases for CTA, i.e. maps, spectrum, lightcurve, in a few variants, e.g. per-run joint, or stacked, or with a simple model (isolated source) and a complex region (a few sources with many fit parameters).

Concretely the tasks here (via a series of small pull requests) are:
- fix the one existing script for a map data reduction (see current tutorials)
- add a few more benchmarks, e.g. for spectrum data reduction, and model fitting or spectral points
- only once that is done and someone is working on this task, are comments welcome about the matrix of analyses to run!
- if we find any issues with unexpectedly high CPU or memory usage, we should go and fix Gammapy
- there's probably several low-hanging fruit for optimisation in Gammapy, such as using float32 instead of float64, or optimising 2-3 functions where all the time is spent. Such optimisations are also really only welcome once it has been established by a real use case benchmark that this is something we care about, so please if you care about optimisation, start here.

This is a good task for new contributors, please get involved!
This might or might not lead to a section / paragraph in the Gammapy paper in fall 2019.

Anyone interested?

**Comment by cdeil** [comment]
@yuwunwong and @#3kaoriinakashima - Thank you for adding benchmarks!

Here's what we have now:

- [analysis_3d.py](https://github.com/gammapy/gammapy-benchmarks/blob/master/benchmarks/analysis_3d.py) - [PNG](https://github.com/gammapy/gammapy-benchmarks/blob/master/results/analysis_3d/0.13.dev9353/2019-07-17/results.png)
- [lightcurve_1d.py](https://github.com/gammapy/gammapy-benchmarks/blob/master/benchmarks/lightcurve_1d.py) - [PNG](https://github.com/gammapy/gammapy-benchmarks/blob/master/results/lightcurve_1d/0.13.dev9353/2019-07-17/results.png)
- [maps_3d.py](https://github.com/gammapy/gammapy-benchmarks/blob/master/benchmarks/maps_3d.py) - [PNG](https://github.com/gammapy/gammapy-benchmarks/blob/master/results/maps_3d/0.13.dev9353/2019-07-17/results.png)
- [spectrum_1d.py](https://github.com/gammapy/gammapy-benchmarks/blob/master/benchmarks/spectrum_1d.py) - [PNG](https://github.com/gammapy/gammapy-benchmarks/blob/master/results/spectrum_1d/0.13.dev9353/2019-07-17/results.png)

The results are as expected: we run at 100% CPU for a single process and OK memory usage (accumulating for spectrum and lightcurve, but overall that doesn't need a lot of RAM, so OK).

More additions / PRs welcome any time. Especially for 3D lightcurve it could be nice to have a benchmark as soon as it's in master, to see where we land in terms of RAM for 10 or 100 runs (cc @registerrier).

**Comment by cdeil** [comment]
I still think this is a very important task. If anyone is interested to work on this for v0.15, please self-assign or leave a comment. Otherwise I'll do it.

The next steps are:
- Fix existing benchmark scripts to latest Gammapy code and re-run.
- Add 3D lightcurve benchmark, following the [light_curve.ipynb](https://docs.gammapy.org/0.14/notebooks/light_curve.html#Make-the-map-datasets) example. It should be super-slow and eat up a ton of RAM.
- Change most benchmarks to use the new high-level interface, i.e. they should have as little code to maintain as possible, yet cover the same main use cases for Gammapy.

I also think it would be nice to run the tutorial notebooks and measure execution time and peak memory usage (see #1425) and record this in a results file in the benchmark repo. (cc @Bultako).

Then we have established the current performance for our major use cases, and we can work a bit to improve it in the coming months. This is just one idea: #2374 There are probably many more low-hanging fruit where we can improve a lot with only some code changes. But if we don't have it measured, creating and reviewing performance-related PRs will be very inefficient.

**Comment by registerrier** [comment]
A comment about implementation here, the loop is opening 100 times the same observation, this might underestimate memory usage and time required for IO, no?

**Comment by cdeil** [comment]
> A comment about implementation here, the loop is opening 100 times the same observation, this might underestimate memory usage and time required for IO, no?

No, the way Gammapy works it re-loads 100 times. The benchmark scripts could be changed to write 100 simulated datasets to disk and load from those. It would just use more disk space, but performance should be exactly the same.

**Comment by cdeil** [comment]
@AtreyeeS - could you please add a README to https://github.com/gammapy/gammapy-benchmarks/tree/master/benchmarks with a bullet list of benchmarks, linking to the existing Python scripts by filename and a few word description for each?
I mean something equivalent like what we have for validation outline [here](https://github.com/gammapy/gammapy-benchmarks/tree/master/validation#analyses).

**Comment by cdeil** [comment]
With #2524 merged, coordinate caching should be fixed now, and memory should behave in a predictable way. @AtreyeeS - Can you please re-run all benchmarks and commit results to check?

**Comment by AtreyeeS** [comment]
Benchmarks run nightly now - closing this issue
